sphinx_needs/schema/config.py

```diff
@@ -0,0 +1,439 @@
+from __future__ import annotations
+
+from enum import Enum, IntEnum
+from typing import Final, Literal, TypedDict
+
+from typing_extensions import NotRequired
+
+EXTRA_OPTION_BASE_TYPES_STR: Final[set[str]] = {
+    "string",
+    "integer",
+    "number",
+    "boolean",
+    "array",
+}
+"""Extra option base types as string that are supported in the schemas."""
+
+EXTRA_OPTION_BASE_TYPES_TYPE = Literal[
+    "string", "integer", "number", "boolean", "array"
+]
+"""Extra option base types as types that are supported in the schemas."""
+
+MAX_NESTED_NETWORK_VALIDATION_LEVELS: Final[int] = 4
+"""Maximum number of nested network validation levels."""
+
+
+class ExtraOptionStringSchemaType(TypedDict):
+    # string is the default type and injected automatically
+    type: NotRequired[Literal["string"]]
+    """Extra option string type."""
+    minLength: NotRequired[int]
+    """Minimum length of the string."""
+    maxLength: NotRequired[int]
+    """Maximum length of the string."""
+    pattern: NotRequired[str]
+    """A regex pattern to validate against."""
+    format: NotRequired[
+        Literal[
+            "date-time",
+            "date",
+            "time",
+            "duration",
+            "email",
+            "uri",
+            "uuid",
+        ]
+    ]
+    """A format string to validate against, e.g. 'date-time'."""
+    enum: NotRequired[list[str]]
+    """A list of allowed values for the string."""
+    const: NotRequired[str]
+    """A constant value that the string must match."""
+    default: NotRequired[str]
+    """Default value used in IDE scenarios for autocompletion."""
+
+
+class ExtraOptionNumberSchemaType(TypedDict):
+    """
+    Float extra option schema.
+
+    The option will still be stored as a string in Sphinx-Needs,
+    but during schema validation, the value will be coerced to the given type.
```

Comment:
This is bad though, the values should be coerced to these from the start, if read from a (string) directive, or just checked if read from a `needs.json`.
This will definitely be the case in ubcode.
What is the path to change sphinx-needs, so that it can handle storing the data as the correct type?

Reply:
While I agree with you that the internal types should change with the definition, my plan was to do that in a 2-step approach. Users should adopt the ontology in a non-breaking way, then we change the internal representation which affects at least filter functions, and data/schema in exported needs.json. My current feeling is changing the types is a bigger thing in the codebase that delays this PR. The adoption of the new interface and the schema feedback is already great for end users. We can switch the type system while adoption happens.